cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -111,9 +111,26 @@ static inline Result<csv::ConvertOptions> GetConvertOptions(
 
   if (!scan_options) return convert_options;
 
-  auto materialized = scan_options->MaterializedFields();
-  std::unordered_set<std::string> materialized_fields(materialized.begin(),
-                                                      materialized.end());
+  auto field_refs = scan_options->MaterializedFields();
+  std::unordered_set<std::string> materialized_fields;
+  materialized_fields.reserve(field_refs.size());
+  // Preprocess field refs. We try to avoid FieldRef::GetFoo here since that's
+  // quadratic (and this is significant overhead with 1000+ columns)
+  for (const auto& ref : field_refs) {
+    if (const auto* name = ref.name()) {
+      // Common case
+      materialized_fields.emplace(*name);
+      continue;
+    }
+    // CSV doesn't really support nested types so do our best
```

Comment:
What is "do our best" mean here?  It seems like we are looking for a column with the same name as the parent field and then trying to cast from a primitive type to the nested type?  Will that ever succeed?

Reply:
This is the same behavior as IPC/ORC: if a nested field is selected, load the entire top-level field and let the projection take care of selecting the child field later. By "do our best" I just mean that CSV doesn't support reading any nested types in the first place so there's nothing to test/this branch should never really get hit, but I'll clarify what's going on here.